extension/chrome/elements/compose-modules/compose-storage-module.ts

```diff
@@ -81,19 +81,47 @@ export class ComposeStorageModule extends ViewModule<ComposeView> {
+    //    more than one public key per recipient. We are blindly getting the first public key,
+    //    and making decisions based on that. We should be using the whole array instead.
     const [storedContact] = await ContactStore.get(undefined, [email]);
     if (storedContact && storedContact.hasPgp && storedContact.pubkey && !storedContact.revoked) {
-      // Potentially check if pubkey was updated - async. By the time user finishes composing, newer version would have been updated in db.
-      // If sender didn't pull a particular pubkey for a long time and it has since expired, but there actually is a newer version on attester, this may unnecessarily show "bad pubkey",
-      //      -> until next time user tries to pull it. This could be fixed by attempting to fix up the rendered recipient inside the async function below.
+      // checks if pubkey was updated, asynchronously. By the time user finishes composing,
+      //    newer version would have been updated in db.
+      // This implementation is imperfect in that, if sender didn't pull a particular pubkey
+      //    for a long time and the local pubkey has since expired, and there actually is a
+      //    newer version available on external key server, this may unnecessarily show "bad pubkey",
+      //    until next time user tries to enter recipient in the field again, which will at that point
+      //    get the updated key from db. This could be fixed by:
+      //      - either life fixing the UI after this call finishes, or
+      //      - making this call below synchronous and using the result directly
+      //
+      // note by Tom 2021-08-10: I wonder if this functionality could be deprecated, it appears that
+      //    `lookupPubkeyFromKeyserversAndUpsertDb` below could cover this usecase, too.
       this.checkKeyserverForNewerVersionOfKnownPubkeyIfNeeded(storedContact).catch(Catch.reportErr);
       return storedContact;
     }
-    return await this.ksLookupUnknownContactPubAndSaveToDb(email, name, storedContact);
+    return await this.lookupPubkeyFromKeyserversAndUpsertDb(email, name, storedContact);
   }
 
-  public ksLookupUnknownContactPubAndSaveToDb = async (email: string, name: string | undefined, existingContact: Contact | undefined): Promise<Contact | "fail"> => {
+  /**
+   * We are searching recipient public key by email every time we enter the recipient.
+   * This is regardless if we already have the public key stored locally or not.
+   * We process the response and if there are new public keys, we save them. If there are
+   *    newer versions of public keys we already have (compared by fingerprint), then we
+   *    update the public keys we already have.
+   * Finally, we return the updated public key back.
+   *
+   * Note by Tom 2021-08-10: It looks to me like the return signature should instad
+   *    be `Promise<Contact[]>` since we deal with a list? And the `fail` situations
```

Comment:
`Promise<Contact>` is actually used for subsequent rendering in `renderPubkeyResult`. If we want to move the "diagnostics" logic to compose-recipients module `evaluateRecipients` method, that's ok, but we don't have to present the results as a list of `Contact`, as this is a bit misleading (`email` field will be copied all over the entries), so the perfect solution would be a special type with `email` field and a list of keys.
Currently, the diagnostics (figuring out the properties of the most appropriate key) is done by the line below.
```
const [preferred] = await ContactStore.get(undefined, [email]);
```
We can, of cause, get rid of the compatibility method `ContactStore.get` and move the key-picking logic to some other place, but we need to find all places where it used.
I see it is used in `GmailElementReplacer.evaluateStandardComposeRecipients()` too

Reply:
> so the perfect solution would be a special type with email field and a list of keys

That sounds good to me

> We can, of cause, get rid of the compatibility method ContactStore.get and move the key-picking logic to some other place, but we need to find all places where it used.
I see it is used in GmailElementReplacer.evaluateStandardComposeRecipients() too

Got it. Here's the code:

```ts
      // pick first usableForEncryption
      let selected = contactWithAllPubkeys.sortedPubkeys.find(entry => !entry.revoked && entry.pubkey.usableForEncryption);
      if (!selected) {
        selected = contactWithAllPubkeys.sortedPubkeys.find(entry => !entry.revoked && entry.pubkey.usableForEncryptionButExpired);
      }
      if (!selected) {
        selected = contactWithAllPubkeys.sortedPubkeys[0];
      }
```

`GmailElementReplacer.evaluateStandardComposeRecipients()` only needs to know if there is any usable key or not. So it could be as simple as:

```diff
-const [contact] = await ContactStore.get(undefined, [email]);
+const pubkeys = await ContactStore.getOneWithAllPubkeys(undefined, [email]);
-if (contact && contact.pubkey) {
+if (pubkeys.filter(p => p.usableForEncryption).length) {
```

Therefore, the actual nuanced logic that relates to choosing public keys for encryption can be actually in the compose modules somewhere. Then we can always use `.getOneWithAllPubkeys`, and never `.get`, which should produce fewer quirks in code and in behavior, I think.

Reply:
`Contact.hasPgp` just shows that a key exists, it may be revoked or expired, but generally I don't mind against moving key-selection logic out of ContactStore. It was implemented to mimick the original ContactStore with the intent of not making too much changes at once. Now that the code is looking stable, we can move on with further refactoring.

Reply:
filed https://github.com/FlowCrypt/flowcrypt-browser/issues/3905